### ISSUE TITLE
Documentation support for shortcodes

### ISSUE THREAD

**Issue by khayyamsaleem** [gohugoio/hugo#7368]
Since shortcodes implemented by the maintainer(s) of a Hugo site could grow a lot as the needs of the site grow, it would be nice if the `hugo` CLI supported a documentation system for shortcodes. The proposal is for shortcode templates to support a "meta" item, similar to how versions are currently supported in shortcodes as a meta tag. 

Then, if one were to run `hugo list shortcodes` or `hugo describe shortcodes`, all the shortcodes available in the site could be printed onto the console, with the title being the name of the shortcode and the description provided by the metadata.

This would have to support both the shortcodes implemented at the repo root level, as well as shortcodes implemented in themes / modules.

**Comment by khayyamsaleem** [comment]
This would leave opportunity to offer a similar documentation system for layouts, partials, and other "resource types."

**Comment by khayyamsaleem** [comment]
Keep alive for now

**Comment by michaeltlombardi** [comment]
A documentation subsystem to allow/enable quickly reviewing docs for shortcodes would be immensely useful; moreover, once it's possible to surface those docs automatically in the CLI, the _information_ can be reused for a theme reference docs export as a static site.

Right now, the only way I'm able to document shortcodes/etc is in a separate site for the theme which you must know exists and has to be handcrafted.

**Comment by davidsneighbour** [comment]
Put your shortcode into a module, then document that module with a proper README.md, done. 

I don't think (my opinion only) that Hugo should support features of third party developers. I also started doing little automatically included content particles that describe the shortcodes available. You would mount content, layouts and add your params in the params section of config or via data configuration. Then the user adds the module and everything maps in nicely.

What I am saying is that anything that has to run extra will make Hugo slower and is not really part of its "list of tasks". Documentation should be the developer's task. 

Problems to keep your eyes on here would be naming conflicts of the resulting web pages, always try working with some form of namespace (I use `content/dnb/shortcodes/shortcodename/index.md` for instance.

Another thing to find out or document would be the implementation of the pages into some form of navigation. In some of my projects, I solved that by adding a `config/development` configuration for menu items that snap into `main` navigation if it exists. This would show only by `hugo server` and not in production.
